CertGetCertificateContextProperty

aluferYochay opened the pull request:
Hi

For start I created this pull request copying one function (CertGetCertificateContextProperty) from our code base.
Once this pull request is approved , I hope many more will follow.

Thanks
Yochay

Commit:
CertGetCertificateContextProperty

AArnott commented on src/Crypt32.Shared/Crypt32+CRYPT_PROV_TYPE.cs:
```diff
@@ -0,0 +1,28 @@
+﻿namespace PInvoke
```
Please confirm that you don't get any build warnings. In this case, I expect a stylecop warning because you're missing the copyright notice at the top of the file. The squigglies in VS should help you automatically correct these with little effort on your part.

AArnott commented on src/Crypt32.Shared/Crypt32+CRYPT_KEY_PROV_INFO.cs:
```diff
@@ -0,0 +1,20 @@
+    public partial class Crypt32
+    {
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+        public struct CRYPT_KEY_PROV_INFO
```
[Our policy is to use pinnable structs](https://github.com/AArnott/pinvoke/blob/master/CONTRIBUTING.md#struct-field-types). Can you correct?

AArnott commented on src/Crypt32.Shared/Crypt32.cs:
```diff
@@ -51,5 +51,12 @@ public static partial class Crypt32
+        public static extern bool CertGetCertificateContextProperty(
+                IntPtr pCertContext,
+                CertPropId dwPropId,
+                IntPtr pvData,
```
`pvData` should be typed as `void*`, and `pcbData` should be `int`. Please review [parameter type guidelines](https://github.com/AArnott/pinvoke/blob/master/CONTRIBUTING.md#method-parameter-types)

AArnott commented:
Overall it looks very good. You seem to have taken effort to fit in nicely with the surrounding code. Thank you!

I've left some comments. Bonus points for adding MSDN documentation as xml doc comments to all the types/members. But that's not strictly required.

AArnott commented on src/Crypt32.Shared/Crypt32.cs:
```diff
@@ -51,5 +51,12 @@ public static partial class Crypt32
         [return: MarshalAs(UnmanagedType.Bool)]
         private static extern bool CertCloseStore(IntPtr hCertStore, CertCloseStoreFlags dwFlags = CertCloseStoreFlags.None);
+
+        [DllImport(nameof(Crypt32), CharSet = CharSet.Unicode, SetLastError = true)]
```
No method parameter takes character/string types. So the `CharSet` parameter in your `DllImport` attribute seems superfluous.

vbfox commented on src/Crypt32.Shared/Crypt32+CRYPT_KEY_PROV_INFO.cs:
```diff
@@ -0,0 +1,20 @@
+    using System;
+    using System.Runtime.InteropServices;
+
+    public partial class Crypt32
```
Each partial declaration should contain a standardized xml comment for example : https://github.com/AArnott/pinvoke/blob/master/src/User32.Desktop/User32+SafeHookHandle.cs#L9

Commit:
1. Adding comments
2. Changing string to char* in structs
3. Changing from Int.Ptr to void* in method declaration

AArnott commented on src/Crypt32.Shared/Crypt32+CRYPT_KEY_PROV_INFO.cs:
```diff
@@ -1,19 +1,52 @@
+        /// </summary>
         [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
-        public struct CRYPT_KEY_PROV_INFO
+        public unsafe struct CRYPT_KEY_PROV_INFO
```
Can you add `[OfferIntPtrPropertyAccessors]` to this struct and set the Custom Tool property for the file? This was previously missing from our contributing.md file (sorry!) but I've corrected it so you can read about it in the last couple paragraphs of [this section](https://github.com/AArnott/pinvoke/blob/master/CONTRIBUTING.md#struct-field-types).

This will make your struct consumable by VB.NET programmers.

AArnott commented:
Wow, that looks awesome @aluferYochay. Thanks so much for responding so well to my earlier comments. I left one more, and I see @vbfox left one as well. If you have time to respond to them, great. Otherwise in the next day or two I can respond to them myself and then merge your PR.

As a side node, thank you for helping us to vet our contributing.md document. We clearly were missing a few things so your PR is helping the overall project more than just the addition you're making. :smile:

Commit:
1. Add partial class comments
2. OfferIntPtrPropertyAccessors attribute to CRYPT_KEY_PROV_INFO struct

AArnott commented:
There are still some build warnings that your PR introduces. You can check out your appveyor build results to see them. You can also check that you're 'clean' locally by running our build.cmd/build.ps1 file at the root of our repo.

Commit:
Fix style warnings

aluferYochay commented:
I still have warning but it does not seem related to my changes.
Please let me know if I need to change anything else.

Thanks
Yochay

AArnott commented:
Thanks. It looks good. I don't see any build warnings.

aluferYochay commented:
When my addition will be part of the official package ?

AArnott commented:
We'll probably push another release to nuget.org this week.

But you [don't have to wait for that](https://github.com/AArnott/pinvoke/blob/master/CONTRIBUTING.md#self-service-releases-for-contributors).

AArnott commented:
BTW, a public release is now on nuget.org that includes this change.

aluferYochay commented:
Thanks!
